lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -2144,7 +2144,38 @@ static bool foldCommonMuxOperation(MuxOp mux, Operation *trueOp,
   return false;
 }
 
-LogicalResult MuxOp::canonicalize(MuxOp op, PatternRewriter &rewriter) {
+// If both arguments of the mux are arrays with the same elements, sink the
+// mux and return a uniform array initializing all elements to it.
+static bool foldMuxOfUniformArrays(MuxOp op, PatternRewriter &rewriter) {
+  auto trueVec = op.getTrueValue().getDefiningOp<hw::ArrayCreateOp>();
+  auto falseVec = op.getFalseValue().getDefiningOp<hw::ArrayCreateOp>();
+  if (!trueVec || !falseVec)
+    return false;
+  if (!trueVec.isUniform() || !falseVec.isUniform())
+    return false;
+
+  auto mux = rewriter.create<MuxOp>(op.getLoc(), op.getCond(),
+                                    trueVec.getUniformElement(),
+                                    falseVec.getUniformElement());
+
+  SmallVector<Value> values;
+  for (unsigned i = 0, n = trueVec.getInputs().size(); i < n; ++i)
+    values.push_back(mux);
```

Comment:
Does `SmallVector<Value> values(trueVec.getInputs().size(), mux);` work?

Reply:
Considered that there is `comb.replicate` we might want to introduce `hw.array_repliacte` or something.

Reply:
That would be useful, although there is some overlap with `array_create`. I think for this particular case we could consider specialising array_create of uniform arrays to store the items more efficiently.